kotlin/reflect/src/main/java/com/squareup/moshi/kotlin/reflect/KtTypes.kt

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright (C) 2020 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.moshi.kotlin.reflect
+
+import com.squareup.moshi.internal.Util
+import kotlinx.metadata.Flag
+import kotlinx.metadata.KmClass
+import kotlinx.metadata.KmClassifier
+import kotlinx.metadata.KmClassifier.TypeAlias
+import kotlinx.metadata.KmClassifier.TypeParameter
+import kotlinx.metadata.KmConstructor
+import kotlinx.metadata.KmProperty
+import kotlinx.metadata.KmType
+import kotlinx.metadata.KmValueParameter
+import kotlinx.metadata.isLocal
+import kotlinx.metadata.jvm.signature
+import java.lang.reflect.Constructor
+import java.lang.reflect.Field
+import java.lang.reflect.Method
+import java.lang.reflect.Type
+
+internal val KmType.isNullable: Boolean get() = Flag.Type.IS_NULLABLE(flags)
+
+private fun defaultPrimitiveValue(type: Type): Any? =
+  if (type is Class<*> && type.isPrimitive) {
+    when (type) {
+      Boolean::class.java -> false
+      Char::class.java -> 0.toChar()
+      Byte::class.java -> 0.toByte()
+      Short::class.java -> 0.toShort()
+      Int::class.java -> 0
+      Float::class.java -> 0f
+      Long::class.java -> 0L
+      Double::class.java -> 0.0
+      Void.TYPE -> throw IllegalStateException("Parameter with void type is illegal")
+      else -> throw UnsupportedOperationException("Unknown primitive: $type")
+    }
+  } else null
+
+internal val KmType.canonicalName: String
+  get() {
+    return buildString {
+      val classifierString = when (val cl = classifier) {
+        is KmClassifier.Class -> createClassName(cl.name)
+        is TypeAlias -> createClassName(cl.name)
+        is TypeParameter -> arguments[cl.id].type?.canonicalName ?: "TypeVar(${cl.id})"
+      }
+      append(classifierString)
+
+      val args = arguments.joinToString(", ") {
+        "${it.variance?.name} ${it.type?.canonicalName ?: "*"}"
+      }
+
+      if (args.isNotBlank()) {
+        append('<')
+        append(args)
+        append('>')
+      }
+
+      // TODO not sure if we care about expressing the other type information here
```

Comment:
anything in particular? I’d rather just wait for tests to fail

Reply:
(the TODO seems needlessly ominous)

Reply:
1d1de9a